xarray/tests/test_backends.py

```diff
@@ -2163,11 +2164,13 @@ def test_open_mfdataset(self):
             with create_tmp_file() as tmp2:
                 original.isel(x=slice(5)).to_netcdf(tmp1)
                 original.isel(x=slice(5, 10)).to_netcdf(tmp2)
-                with open_mfdataset([tmp1, tmp2]) as actual:
+                with open_mfdataset([tmp1, tmp2], combine='manual',
+                                    concat_dim='x') as actual:
```

Comment:
why are you updating these tests? Is it just to increase coverage a bit, or will these previously valid calls now be invalid? If so, then this is probably going to need a deprecation cycle. It is not obvious to me why this needs changing.

Reply:
> will these previously valid calls now be invalid?

If we do what you suggested in your other comment then these particular calls will still be valid (so I would revert these particular tests).

> why are you updating these tests?

Because as proposed right now this is going to be a breaking change :confused:. 

Due to what me and @shoyer decided during the discussion [here](https://github.com/pydata/xarray/pull/2553#issuecomment-445419876), then some previously-valid inputs to `auto_combine` (and hence to `open_mfdataset`) will become invalid.

As far as I know then there is only one type of input which will be go from valid to invalid (assuming we implement your comment). That's the case of wanting to combine a set of datasets which need to be both concatenated and merged, but you want to concatenate along a dimension which does not have a corresponding coordinate variable. The old `auto_combine` would have just concatenated them in the order they were given, but the new one will complain that it doesn't know what the order should be without a global coordinate to order them. The new `manual_combine` won't do it because you need to perform two operations (concat then merge), which currently requires a two-dimensional list input, i.e:

```python
datasets = [[var1_x0, var1_x1], [var2_x0, var2_x1]]
result = manual_combine(datasets, concat_dim=['x', None])
```

This is why I wrote in the what's new that it's a potentially breaking change - whether or not it justifies a deprecation cycle I guess is not up to me....

It might be possible to make `manual_combine` more flexible so that it would accept

```python
datasets = [var1_x0, var1_x1, var2_x0, var2_x1]
result = manual_combine(datasets, concat_dim=['x', None])
```

(which currently will throw an error because `len(concat_dim) =/= datasets.ndims` basically), but I'm not sure this solution is any clearer.

@shoyer what do you think?

Reply:
We can (and probably should) deprecate this, but we try really hard to avoid hard backwards compatibility breaks.

Could we implement some fall-back logic to do this? e.g., by explicitly checking for the scenario where only a single dimension to concatenate is provided, but the datasets also need to be merged. In this case we would append another dimension `None` and issue a warning that this behavior won't be supported in the future.

Reply:
@shoyer I've changed the API to how I think it should be finally (e.g. removed automatic inferring of `concat_dim` in `manual_combine`as discussed above), and tried to make the documentation as clear as possible. However I don't know what you want to do about the backwards-compatibility issue.

The problem is that with separate `auto_combine` and `manual_combine` functions (which I think make sense to have), then the new `auto_combine`'s behaviour is so different to the old one that creating full backwards compatibility will basically involve having both the new and the old one present in the code simultaneously. This is feasible, but also the new and old `auto_combine` will give different results for the same input, and you can't know whether a user wants the new or old version to be used. Therefore the only way I can see to slowly phase the old one out is to have some kind of default `OLD=True` flag or something. That would probably be confusing documentation-wise though.

Another idea would be to just change it so the current behaviour has a loud warning saying the default behaviour will change soon, and once it changes then raise a warning whatever you do to get people to check that their datasets are still be combined in the way they want them to.

Reply:
To recap, there are two cases whose behavior we want to change:
1. 1D lists, which require both concatenation *and* merging and for which the concentration dimension was given explicitly
2. 1D lists, which should be concentrated in order of appearance, not coordinate order.

In both cases, we eventually want to change the default behavior of `xarray.open_mfdataset(paths)`, and there's an easy way to clearly opt-in to the old behavior even with the new code (because these arguments weren't valid previously):
1. `xarray.open_mfdataset(paths, concat_dim=['time', None])`
2. `xarray.open_mfdataset(paths, combine='manual')`

And in both cases, it would also be straightforward to opt-in to the new behavior early by writing `combine='auto'`.

So I think a deprecation cycle would be relatively feasible. The logic goes roughly like this:
- If a user makes use of any new arguments that weren't previously valid, opt-in to the new behavior. Something this means setting a default value for a new argument to `None`, so we can detect whether it was explicitly provided or not.
- Otherwise, check to see behavior would change for this usage in the future. If xarray's behavior would indeed change, issue a `FutureWarning` that suggests two alternative syntax options:
   - to maintain the current behavior in a backwards compatible way, or
   - to opt-in to the new behavior early

> The problem is that with separate `auto_combine` and `manual_combine` functions (which I think make sense to have), then the new `auto_combine`'s behaviour is so different to the old one that creating full backwards compatibility will basically involve having both the new and the old one present in the code simultaneously.

Yep, this is sometimes what we have to do.

Reply:
I'm trying to implement this now, but I've already found one case I don't know how to deal with:

`xarray.auto_combine([1d_list])` should be concatenated in order of appearance in the old version, but in coordinate order in the new version. However there's no way to know which one the user wants, and no way to give them the option to opt straight-in to the new behaviour, because the argument signatures are identical for the old and the new version in this case. To opt-in to the new way we would have to introduce a new argument just for that purpose, which would itself then be deprecated in the next version.

Reply:
For `open_mfdataset`, we can use the trick of setting `combine=None` as a default. Then:
- `combine=None`: backwards compatible behavior, raising warnings as needed
- `combine='auto'`: opt-in to automatic combination
- `combine='manual'`: stick with manual combination

Reply:
Okay, but then where do I put the description of what the new `auto_combine` does? And that means it would be possible to opt-in to the new `auto_combine` behaviour from `open_mfdataset`, but not from `auto_combine`!